src/dns.rs

```diff
@@ -0,0 +1,97 @@
+mod local;
+mod default;
+mod cache;
+
+use crate::types::Result;
+use std::net::IpAddr;
+use std::time;
+
+/// A single IP entry in the DNS cache. Each ip entry has an expiration time (based on the TTL)
+/// and a last-used timestamp. When the entry is expired, the entry is found to be stale and should
+/// be refreshed by the resolver. When expired == 0, the entry is considered to be a local override
+/// and should not be refreshed by the resolver.
+/// @todo: Does that information belong here? Or should it be in the cache layer?
+struct IpEntry {
+    /// Actual IP address (either ipv4 or ipv6)
+    ip: IpAddr,
+    /// Epoch when ip address expires
+    expired: u64,
+    /// Epoch when this entry was last used
+    last_used: u64,
+}
+
+/// A DNS entry is a simple domain name to IP address mapping
+#[derive(Default)]
+struct DNSEntry {
+    // domain name
+    domain_name: String,
+    // List of IPv4 addresses for this domain name
+    ipv4: Vec<IpEntry>,
+    // List of IPv6 addresses for this domain name
+    ipv6: Vec<IpEntry>,
+    // Round robin pointer in case we need to round robin
+    rr_ipv4_ptr: usize,
+    // Round robin pointer in case we need to round robin
+    rr_ipv6_ptr: usize,
+}
+
+impl DNSEntry {
+    /// Instantiate a new domain name entry
+    pub(crate) fn new(domain: &str) -> DNSEntry {
+        DNSEntry {
+            domain_name: domain.to_string(),
+            ..Default::default()
+        }
+    }
+}
+
+impl DNSEntry {
+    /// Returns true when this entry has a ipv4 address
+    fn has_ipv4(&self) -> bool {
+        !self.ipv4.is_empty()
+    }
+
+    /// Returns true when this entry has a ipv6 address
+    fn has_ipv6(&self) -> bool {
+        !self.ipv6.is_empty()
+    }
+
+    /// Retrieves the next ipv4 address in the round robin list
+    fn next_ipv4(&mut self) -> IpAddr {
+        let mut entry = self.ipv4.get_mut(self.rr_ipv4_ptr).expect("Invalid round robin pointer");
+        self.rr_ipv4_ptr = (self.rr_ipv4_ptr + 1) % self.ipv4.len();
+
+        // Update the last used timestamp
+        entry.last_used = time::Instant::now().as_secs();
+
+        entry.ip
+    }
+
+    /// Retrieves the next ipv6 address in the round robin list
+    fn next_ipv6(&mut self) -> IpAddr {
+        let mut entry = self.ipv6.get_mut(self.rr_ipv6_ptr).expect("Invalid round robin pointer");
+        self.rr_ipv6_ptr = (self.rr_ipv6_ptr + 1) % self.ipv6.len();
+
+        // Update the last used timestamp
+        entry.last_used = time::Instant::now().as_secs();
+
+        entry.ip
+    }
+}
+
+/// Type of DNS resolution
+#[derive(Debug, PartialEq, Clone)]
+pub enum ResolveType {
```

Comment:
I don't know exactly who will be in charge of this, but somewhere we must decide if we resolve the ipv4 or ipv6 of a domain. This is the ResolveType we can add to the resolve() function.